Config change at runtime

prognant opened the pull request:
### What does this PR do?

Add command to change the log level at runtime.

### Motivation
The main reason was to be able to change log_level at runtime to ease troubleshooting.

### Additional Notes
Partly based on https://github.com/DataDog/datadog-agent/pull/2829

prognant committed:
typo + log level sanity check

prognant committed:
Add runtime configurable settings facility - API+CLI

Implement a generic way to add runtimg configurable setting.
It include the ability to change log_level at runtime using
the aforementioned work.

prognant committed:
Add relevant release note

hush-hush commented on cmd/agent/app/config.go:
```diff
@@ -78,3 +97,65 @@ func requestConfig() (string, error) {
+	if err != nil {
+		return err
+	}
+	url := fmt.Sprintf("https://%v:%v/agent/config/list-runtime", ipcAddress, config.Datadog.GetInt("cmd_port"))
```
Could we set `/agent/config/list-runtime` as a var ?

prognant replied:
I agree with @remeh, I think it is unlikely that this route will serve elsewhere.

hush-hush replied:
moving it to a var is the first step to make sure we won't duplicate it anywhere else.

hush-hush commented on pkg/config/runtime_setting.go:
```diff
@@ -0,0 +1,68 @@
+	registerRuntimeSetting(RuntimeSetting(ll))
+}
+
+// LogLevel wraps operations to change log level at runtime
```
Could we move this to it's own file `runtime_setting_log_level.go` or something ?

remeh replied:
+1
We should have more runtime settings in the future and the `runtime_setting` part in `runtime_setting_log_level.go` definitely looks like a namespace to me, that's why I would even consider the runtime setting part having its proper sub-package in the `config` one. Feel free to ignore and only move it to a file like Maxime said if you prefer, we can move it later in another package when the needs will arise.

prognant replied:
I hesitated given the fact that log level implementation was rather small, but I agree it make sens to put it in a separate file.

hush-hush commented on releasenotes/notes/runtime-config-change-06132b1d9b1a2bd9.yaml:
```diff
@@ -0,0 +1,3 @@
+features:
+  - |
+    Add the ability to change log_level at runtime.
```
We should explain a bit more how or point to the official doc we if we have one. So users reading the CHANGELOG knows how to use it.

prognant replied:
Indeed the release note would benefit from being a little bit more verbose.

hush-hush commented on cmd/agent/api/agent/agent.go:
```diff
@@ -278,6 +281,40 @@ func getRuntimeConfig(w http.ResponseWriter, r *http.Request) {
+		body, _ := json.Marshal(map[string]string{"error": err.Error()})
+		http.Error(w, string(body), 500)
+		return
+	}
```
I think the logic of setting a runtime settings should be part of the config package. The rest of the code should not have to be aware how runtime settings are stored for example.
Either `config.SetRuntimeSettings` or make runtimeSettings it's own package ?

prognant replied:
Having method to set / get runtime settings is good idea, however I think it makes sense to let this piece of code in the config package.

hush-hush commented on cmd/agent/api/agent/agent.go:
```diff
@@ -278,6 +281,40 @@ func getRuntimeConfig(w http.ResponseWriter, r *http.Request) {
+	configurableSettings := make(map[string]string)
+	for name, setting := range config.RuntimeSettings {
+		configurableSettings[name] = setting.Description()
+	}
```
Same here, we should have a function to get those: `config.getRuntimeSettings() -> map[string]string` ?

prognant replied:
Will do, cf. above, it will better isolate the runtime settings stuff inside the config package.

remeh commented on pkg/util/log/log.go:
```diff
@@ -556,3 +556,12 @@ func GetLogLevel() (seelog.LogLevel, error) {
 	return seelog.InfoLvl, errors.New("cannot get loglevel: logger not initialized")
 }
+
+// ChangeLogLevel change the current log level, valide levels are trace, debug, info, warn, error, critical and off
```
nit: ChangeLogLevel changes

prognant replied:
Will correct.

remeh commented on cmd/agent/api/agent/agent.go:
```diff
@@ -278,6 +281,40 @@ func getRuntimeConfig(w http.ResponseWriter, r *http.Request) {
+
+	if _, ok := config.RuntimeSettings[setting]; !ok {
+		body, _ := json.Marshal(map[string]string{"error": "unknown setting"})
+		http.Error(w, string(body), 500)
```
A `400` would be better here.

prognant replied:
I agree a 400 is a better match for this error.

remeh commented on cmd/agent/api/agent/agent.go:
```diff
@@ -278,6 +281,40 @@ func getRuntimeConfig(w http.ResponseWriter, r *http.Request) {
+	}
+	if err := config.RuntimeSettings[setting].Set(value); err != nil {
+		body, _ := json.Marshal(map[string]string{"error": err.Error()})
+		http.Error(w, string(body), 500)
```
Ditto.

remeh replied:
Actually, nope. 500 is ok here.

prognant replied:
If an error occurs here, the setting exist but changing it failed, thus I think returning a 500 code is correct.

remeh commented on cmd/agent/app/config.go:
```diff
@@ -78,3 +97,65 @@ func requestConfig() (string, error) {
+	}
+	fmt.Println("=== Settings that can be changed at runtime ===")
+	for setting, desc := range settings {
+		fmt.Printf("%s: %s\n", setting, desc)
```
What do you think of using something like: `%s\t\t\t%s\n`. Feel free to ignore.

prognant replied:
I agree, larger space look nicer.

remeh commented on cmd/agent/app/config.go:
```diff
@@ -18,42 +20,59 @@ import (
+		Long:  ``,
+		RunE:  listRuntimeConfigurableValue,
+	}
+	setCommand = &cobra.Command{
```
Maybe a `get <setting>` method could be useful, what do you think?

prognant replied:
I agree, but as per the the design discussion I believed that list-runtime was deemed to be enough, anyway it can be added relatively easily, but I'm not 100% sure it is very useful.

remeh replied:
I was thinking about the other Agents which are getting some info from the infra agent, calling it with some commands (for instance the hostname). There is a RFC in the pipe to do this better, but because a `get` is easy to implement that's probably worth (avoid some parsing logic)? We could discuss this with the team if you want to.

prognant replied:
I implemented a `get` method.

remeh reviewed (changes requested):
Looks good overall! 👍 
I made a few comments needing small changes and that would be awesome if you can update the PR description with a light manual on how to use this new feature? This way while looking on Github history we'll have more info. I agree with Maxime about the CHANGELOG desc as well.
Thank you!

prognant committed:
isolate log level runtime settings

prognant committed:
isolate runtime configuration methode inside config package

prognant commented on pkg/config/runtime_setting.go:
```diff
@@ -0,0 +1,68 @@
+)
+
+// RuntimeSettings registers all runtime editable config
+var RuntimeSettings = make(map[string](RuntimeSetting))
```
Will do.

prognant committed:
Address review comments

prognant commented:
Found an issue with the current implementation, when the agent start with log_level=off, changing the log_level at runtime does not work but no error occurs during the change. As soon as a fix is available it will be added to the PR.

prognant committed:
Add details to the release note

prognant committed:
Remove min log level from seelog config

The log level logic is implemented inside
the agent code to allow dynamic change at runtime

prognant committed:
Add a get <setting> command for consistency

prognant commented on pkg/config/log.go:
```diff
@@ -97,13 +101,12 @@ func SetupLogger(loggerName LoggerName, logLevel, logFile, syslogURI string, sys
 	if jsonFormat {
 		formatID = "json"
 	}
 
```
I remove the minlevel because :
* The should-I-log-based-on-current-log_level logic is already implemented in the agent (cf. https://github.com/DataDog/datadog-agent/blob/master/pkg/util/log/log.go#L103-L109, and caller like https://github.com/DataDog/datadog-agent/blob/master/pkg/util/log/log.go#L371 that are already evaluated for each logging call)
* It was the most convenient way to enable log_level increase at runtime (and it does not come with any performance penalty because we already test if a log message is eligible for logging (or not) given its log level)

remeh reviewed (commented):
Some last comments!

prognant committed:
typo

prognant commented on releasenotes/notes/runtime-config-change-06132b1d9b1a2bd9.yaml:
```diff
@@ -0,0 +1,10 @@
+    the log_level to debug the following command should be used:
+    ``agent config set log_level debug``, all runtime-configurable
+    settings can be listed using ``agent config list-runtime``. The
+    log_level may also be fetch using the ``agent config get log_level``
```
Corrected.

prognant commented on pkg/config/runtime_setting_log_level.go:
```diff
@@ -0,0 +1,41 @@
+// LogLevel wraps operations to change log level at runtime
+type logLevel string
+
+var ll logLevel = "log_level"
```
I agree, is at least as readable and shorter, so it's better, will do.

prognant commented on pkg/config/runtime_setting_log_level.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// LogLevel wraps operations to change log level at runtime
```
It's good idea to make it less ambiguous. Will do.

prognant committed:
Address remarks

remeh reviewed (dismissed):
:shipit:

hush-hush commented on pkg/config/runtime_setting.go:
```diff
@@ -0,0 +1,72 @@
+/// Unless explicitly stated otherwise all files in this repository are licensed
```
Could we had tests for this and runtime_setting_log_level.go ?

prognant replied:
done

prognant committed:
moving url paths to var

prognant committed:
add unit tests
